Improve drag and drop into array property editors

ttencate opened the pull request:
This makes a couple of things work that you'd expect to work:

- Dragging a node now highlights properties accepting that node's type as valid drop targets. The drop was already possible, it just wasn't indicated like it was for other types of properties.
- Custom resources can now be dropped onto array property editors that accept that resource.
- Nodes with a script can now be dropped onto array property editors that accept that type of script.
- Items can be added to an array property by dropping them on the `+ Add Element` button. The button is highlighted to indicate this.

Further details and rationale for implementation can be found in the individual commit messages.

Fixes https://github.com/godotengine/godot-proposals/issues/6862, which currently has 33 upvotes and no objections.

CC @YuriSizov who brought this up on Mastodon.

Demo video:

https://github.com/user-attachments/assets/d61030cc-49f7-4c5a-bef1-79db08e1cf90

Test project:

[dragdroptest.zip](https://github.com/user-attachments/files/18708228/dragdroptest.zip)

ttencate committed:
Highlight Node and NodePath properties as valid drop targets

This was already done for Resource properties; this just makes things
consistent.

ttencate committed:
Allow dragging custom resources onto array property editor

This duplicates some of the logic in EditorResourcePicker, but that's
unavoidable: EditorResourcePicker works on a single, loaded resource,
whereas we'd like this to be efficient and not need to load all
(potentially many) dragged resources. Instead, we use the
EditorFileSystem cache to get the information we need.

Note that EditorResourcePicker also supports some automatic conversions,
such as from Shader to ShaderMaterial. This change does not attempt to
make that work for arrays as well.

ttencate committed:
Allow dragging custom node types onto array property editor

For example, an Array[MyScript] property now accepts any node with the
script MyScript on it. This works for Node and NodePath arrays alike,
and also with the @export_node_path annotation.

ttencate committed:
Allow dragging resources (not just files) onto array property editor

The most common scenario is dragging a file from the filesystem dock,
which drags it as type "file", not "resource". This already worked. This
change also makes it possible to drag a resource from another property
editor.

ttencate committed:
Allow dragging items onto array property editor Add Element button

Instead of having to first click the button, then drag the item into the
new slot, this allows just dragging the item onto the button directly.
The button is highlighted as a valid drop target to signal this.

KoBeWi commented:
Custom Resource Array gets highlighted for CustomResource.gd, even though it doesn't (and shouldn't) accept it.

https://github.com/user-attachments/assets/685b7178-160c-4506-bd94-3ef4930b234c

ttencate commented:
@KoBeWi Great find! But that bug is also present in the current master, not caused by this PR :grin: In fact, you can also drag `CustomResource.gd` onto the `custom_resource` property, which even accepts it despite being the wrong type. (You can also drag it onto `resource` and `resource_array`, but that makes sense because `Script` inherits from `Resource`.)

I'll see if I can find out why that happens and contribute a fix, but I don't think it needs to block this PR.

ttencate commented:
Correction: it _does_ need some work in this PR because I copied the faulty logic into `EditorPropertyArray` for the case of dropping a `"resource"` type (e.g. from another property) onto the array itself.

The cause is this bit here:

https://github.com/godotengine/godot/blob/36d90c73a843afa2807a0b8dcbfbf52bdb8a759c/editor/editor_resource_picker.cpp#L689-L692

I'm not sure what the purpose of that code is. `_is_type_valid` already checked whether the type inherits from one of the allowed types, both for built-in and custom classes.

`EditorNode::get_object_custom_type_name()` returns the `global_class` name of the script set on the object, but also if the object is _itself_ a `Script`, [it returns its `global_class`](https://github.com/godotengine/godot/blob/36d90c73a843afa2807a0b8dcbfbf52bdb8a759c/editor/editor_node.cpp#L4790-L4792). No idea why that is useful; it seems like conflating two different things. This method was introduced in 4f721788681 for use in a tooltip in `SceneTreeEditor`, where it's only called for `Node`s so the object can never be a `Script` anyway. Since then, usage has been added in e8f15f79968 for the `EditorResourcePicker`, where it's explicitly only used to get the `global_class` of `Script` objects directly. And this was subsequently changed in d32f2700ffd, now calling it with the resource itself rather than the resource's script, introducing the bug.

For now, I'll add another commit to this PR that fixes the bug in the least invasive way possible, and also modify the commit where I copied the bug to the `EditorPropertyArray`.

ttencate committed:
Fix resource property editor incorrectly accepting script files

See discussion on #102534. Before this fix, it was possible to drag a
file `CustomResource.gd` with `class_name CustomResource` onto a
property of type `CustomResource`. Of course, the intention is that only
`Resource`s with the `CustomResource` script attached are accepted.

(Actually accepting the script, but creating a resource from it
automatically, is left as a future improvement.)

ttencate commented:
Should all be fixed now, please take another look.

FYI, I would rather not squash these commits together, because each of them can stand on its own. If any regressions are introduced, bisecting and focused reverting is easier this way.

KoBeWi commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -551,6 +554,13 @@ void EditorPropertyArray::_button_draw() {
 }
 
+void EditorPropertyArray::_button_add_item_draw() {
+	if (button_add_item && dropping) {
```
`button_add_item` always exists here, no? It's where the `draw` signal comes from.

ttencate replied:
True, this is just defensive programming. Want me to remove it?

KoBeWi commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -587,6 +603,28 @@ bool EditorPropertyArray::_is_drop_valid(const Dictionary &p_drag_data) const {
+			script_class = EditorNode::get_singleton()->get_object_custom_type_name(res->get_script());
+		}
+
+		for (int j = 0; j < allowed_type.get_slice_count(","); j++) {
```
This loop 😬
I know it's copied, but you should do `split()` and iterate the string array instead.

KoBeWi commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -587,6 +603,28 @@ bool EditorPropertyArray::_is_drop_valid(const Dictionary &p_drag_data) const {
+			script_class = EditorNode::get_singleton()->get_object_custom_type_name(res->get_script());
+		}
+
+		for (int j = 0; j < allowed_type.get_slice_count(","); j++) {
```
I noticed some very bad loops in the code and opened #102630
You could change this loop to match my changes.

ttencate replied:
I noticed that too, did not want to mess with it for fear of introducing bugs. The bad loops don't matter much, because there's usually just one item in the list anyway.

ttencate replied:
Anyway, fixed up my new addition of the bad pattern. We'll have a merge conflict on the first one though.

ttencate commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -587,6 +603,28 @@ bool EditorPropertyArray::_is_drop_valid(const Dictionary &p_drag_data) const {
 
+	if (drop_type == "resource") {
+		Ref<Resource> res = drag_data["resource"];
+		if (!res.is_valid()) {
```
Done.

ttencate commented:
Rebased onto master, fixed conflicts with b9a60574b41f5ee06454163c628eab06fa3bc926. Now that the dust has settled on the 4.4 release, I guess this is ready to merge?

Repiteo commented:
Needs squash before this can be merged

ttencate commented:
@Repiteo As mentioned [above](https://github.com/godotengine/godot/pull/102534#issuecomment-2644868311):

> FYI, I would rather not squash these commits together, because each of them can stand on its own. If any regressions are introduced, bisecting and focused reverting is easier this way.

Each of them could even have been a separate PR, but the overhead didn't seem worth it.

Still want me to squash?

Repiteo reviewed (approved):
On second thought, nah. The squashing isn't a hard-rule, and doesn't need to be enforced when you've done such a good job at making atomic commits. We'll ship it as-is!

Repiteo commented:
Thanks!

azur-wolve commented:
adding a red tone on fields that cant accept it would be a plus

ttencate commented:
> adding a red tone on fields that cant accept it would be a plus

I think it's better to draw attention to properties that *can* be interacted with, but you can always open a proposal!
